WearTilesKotlin/app/src/main/java/com/example/wear/tiles/messaging/MessagingTileService.kt

```diff
@@ -35,77 +35,87 @@ import kotlinx.coroutines.flow.map
- * uses an Image.
+ * The main function, [tileRequest], is triggered when the system calls for a tile. Resources are
+ * provided with the [resourcesRequest] method, which is triggered when the tile uses an Image.
  */
 class MessagingTileService : CoroutinesTileService() {
-    private lateinit var tileStateFlow: StateFlow<MessagingTileState?>
-    private lateinit var renderer: MessagingTileRenderer
     private lateinit var repo: MessagingRepo
     private lateinit var imageLoader: ImageLoader
+    private lateinit var renderer: MessagingTileRenderer
+    private lateinit var tileStateFlow: StateFlow<MessagingTileState?>
 
     override fun onCreate() {
         super.onCreate()
 
-        val appContainer = (application as TilesApplication).appContainer
-        renderer = appContainer.renderer
-        repo = appContainer.repo
-        imageLoader = ImageLoader(this)
+        repo = MessagingRepo(this)
+        imageLoader = Coil.imageLoader(this)
+        renderer = MessagingTileRenderer(this)
 
         tileStateFlow = repo.getFavoriteContacts()
-            .map {
-                buildState(it)
-            }
+            .map { contacts -> MessagingTileState(contacts) }
             .stateIn(
                 lifecycleScope,
                 started = SharingStarted.WhileSubscribed(5000),
                 initialValue = null
             )
     }
 
-    private fun buildState(contacts: List<Contact>): MessagingTileState {
-        return MessagingTileState(contacts)
-    }
-
+    /**
+     * Read the latest data, delegating to the [MessagingTileRenderer] which creates a layout to
+     * which it binds the state.
+     */
     override suspend fun tileRequest(requestParams: TileRequest): Tile {
-        val tileState = readTileState()
-
+        val tileState: MessagingTileState = latestTileState()
         return renderer.renderTimeline(tileState, requestParams)
     }
 
-    private suspend fun readTileState(): MessagingTileState {
+    /**
+     * Reads the latest state from the flow, and updates the data if there isn't any.
+     */
+    private suspend fun latestTileState(): MessagingTileState {
         var tileState = tileStateFlow.filterNotNull().first()
 
         if (tileState.contacts.isEmpty()) {
-            updateContacts()
+            refreshData()
             tileState = tileStateFlow.filterNotNull().first()
         }
         return tileState
     }
 
-    private suspend fun updateContacts() {
+    /**
+     * If our data source (the repository) is empty/has stale data, this is where we could perform
+     * an update.
+     *
+     * For this sample, we're updating the repository with fake data
+     * ([MessagingRepo.knownContacts]).
+     */
+    private suspend fun refreshData() {
         repo.updateContacts(MessagingRepo.knownContacts)
     }
 
+    /**
+     * Downloads bitmaps from the network and passes them to [MessagingTileRenderer] to add as
+     * image resources (alongside any local resources).
```

Comment:
Should we talk about what happens if image loading fails?  How to mitigate? e.g. ensure you are using the same ImageLoader and cache that the main app uses, avoid hitting the network for these requests because network may not be available?

Reply:
We can come back and update this when we have a fallback strategy (#495), but I don't think it's worth adding that info here because it's not easily demonstrated by the sample; should be in a post/DAC.